### PR TITLE
Add profiling for building and painting.

### PR DESCRIPTION
**Pull request by apwilson**

### CODE REVIEW

**Comment by eseidelGoogle** [comment]
@Hixie or @abarth should comment here.  My understanding is there is a non-zero cost to these timeline calls which has prevented us from having very low-level ones like this in the framework before.  I know @johnmccutchan and/or @rmacnak-google worked to make them cheaper in the past, but I'm not sure the current status.

**Comment by Hixie** [comment]
Yeah, these should at a minimum be in asserts and guarded by a debug flag.

**Comment by apwilson** [comment]
I don't think they should be in asserts - I want to see this info in --profile mode.

**Comment by Hixie** [comment]
I suspect that they would skew the numbers too much to be useful in --profile mode.

**Comment by johnmccutchan** [comment]
If `paintChild` is called in a loop, there will be a measurable overhead. 

We could put it behind a flag that we control over the service protocol something like:

if (timelinePaints) {
  Timeline.startSync(...)
}

...

if (timelinePaints) {
  Timeline.finishSync();
}

**Comment by Hixie** [comment]
paintChild is literally our hottest code path in the entire system. I wouldn't want to even have a branch there in release mode.

**Comment by apwilson** [comment]
I've found these traces to be immensely helpful even if they skew things.  I'm not sure what the solution is here but from my perspective these are vital to any performance debugging.  

What happens to Timeline calls in --release (and not --profile)  are they compiled out?  Can we do something similar to --profile by making these functions have a slightly different signature or something?

**Comment by apwilson** [comment]
Also, the words "skew the numbers" and "measurable overhead" have been mentioned - do we have actual numbers behind this?  In a typical frame I've seen paint and build calls that on a bad frame may add up to 100's to 1000's of calls.  What kind of overhead would I expect from that?

Also, if I'm doing that much building and painting in a frame that's where performance optimizations need to focused anyway to reduce that.

**Comment by apwilson** [comment]
I'll move it to debug with flag for now.

**Comment by Hixie** [comment]
In release mode the overhead is reduced a bit, but also we just don't have many of these per frame right now so it's not been a big deal.

@abarth can talk to the overhead issue, I remember we took some out of the layer code because the timeline work ended up being a double digit proportion of the layer code

**Comment by apwilson** [comment]
@Hixie Is this change good now?

**Comment by johnmccutchan** [comment]
@apwilson putting the Timeline code in the assert() call means it is only active in debug builds.

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -95,12 +95,26 @@ class PaintingContext {
   /// into the layer subtree associated with this painting context. Otherwise,
   /// the child will be painted into the current PictureLayer for this context.
   void paintChild(RenderObject child, Offset offset) {
+    assert(() {
+      if (debugProfilePaintsEnabled) {
+        Timeline.startSync('${child.runtimeType}');
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: nit: flutter style doesn't put braces around one-line if statement bodies

File: packages/flutter/lib/src/widgets/debug.dart [hunk 2]
```diff
@@ -48,6 +48,9 @@ bool debugPrintScheduleBuildForStacks = false;
 /// This can help track down framework bugs relating to the [GlobalKey] logic.
 bool debugPrintGlobalKeyedWidgetLifecycle = false;
 
+/// Adds [Timeline] events for every Widget built.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/debug.dart]: If you could add some docs here on how to actually use that information (e.g. pointing to the observatory timeline UI, or pointing to a doc that talks about how to use timeline events), that would be great.

**Comment by Hixie** [comment]
LGTM.

**Comment by apwilson** [comment]
@johnmccutchan Yeah I know.  It's a compromise.

**Pull request merged**